torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,282 @@ int BestEffortReplay::findFirstMismatchedID(
+      const TensorView* consumer) {
+    // Collect which root axes are in consumer that are not in producer because
+    // of broadcasting
+    std::unordered_set<IterDomain*> consumer_bcast_roots_not_in_producer;
+
+    const auto c2p_root_map =
+        PairwiseRootDomainMap(producer, consumer)
+            .mapConsumerToProducer(consumer->domain(), producer->domain());
+
+    for (auto consumer_root_id : consumer->getRootDomain()) {
+      if (consumer_root_id->isBroadcast()) {
+        if (c2p_root_map.find(consumer_root_id) == c2p_root_map.end()) {
+          consumer_bcast_roots_not_in_producer.emplace(consumer_root_id);
+        }
+      }
+    }
+
+    // We have root axes in consumer that don't exist in producer, now forward
+    // those to include all id's in consumer comprised of only axes not in
+    // producer.
+    auto consumer_bcast_ids_not_in_producer =
+        consumer_bcast_roots_not_in_producer;
+
+    std::vector<Expr*> consumer_history = ExprSort::getExprs(
+        FusionGuard::getCurFusion(),
+        std::vector<Val*>(
+            consumer->domain()->domain().begin(),
+            consumer->domain()->domain().end()));
+
+    auto isValInConsumer =
+        [&consumer_bcast_ids_not_in_producer](Val* input_val) {
+          if (input_val->isA<IterDomain>()) {
+            return consumer_bcast_ids_not_in_producer.find(
+                       input_val->as<IterDomain>()) !=
+                consumer_bcast_ids_not_in_producer.end();
+          } else {
+            return true;
+          }
+        };
+
+    for (auto expr : consumer_history) {
+      // If expr inputs are all in consumer_bcast_ids_not_in_producer, than so
+      // are all outputs
+      if (std::all_of(
+              expr->inputs().begin(), expr->inputs().end(), isValInConsumer)) {
+        // add all outputs to not being in producer
+        for (auto output_ids :
+             ir_utils::filterByType<IterDomain>(expr->outputs())) {
+          consumer_bcast_ids_not_in_producer.emplace(output_ids);
+        }
+      } else if (
+          expr->isA<Merge>() &&
+          std::any_of(
+              expr->inputs().begin(), expr->inputs().end(), isValInConsumer)) {
+        auto merge_expr = expr->as<Merge>();
+        // If
+        // - one of the inputs is made of id's in consumer that don't map to
+        // producer (bcast axes),
+        // - && the other input maps to an id in both consumer and producer
+        // - && this is a merge
+        //   for the sake of BestEffortReplay we can forward the input mapping
+        //   to both consumer and producer to the output of the expression
+        std::vector<IterDomain*> forwarded_ids;
+        std::vector<IterDomain*> compliment_ids;
+
+        for (auto input_id :
+             ir_utils::filterByType<IterDomain>(expr->inputs())) {
+          if (consumer_bcast_ids_not_in_producer.find(input_id) ==
+              consumer_bcast_ids_not_in_producer.end()) {
```

Comment:
Just to confirm, is this equivalent to `!isValInConsumer(input_id)`?

Reply:
Yes, thanks, will reuse it.